src/native/managed/cdacreader/src/Contracts/Thread.cs

```diff
@@ -35,24 +49,77 @@ static IContract IContract.Create(Target target, int version)
 {
     private readonly Target _target;
     private readonly TargetPointer _threadStoreAddr;
+    private readonly ulong _threadLinkOffset;
 
     internal Thread_1(Target target, TargetPointer threadStore)
     {
         _target = target;
         _threadStoreAddr = threadStore;
+
+        // Get the offset into Thread of the SLink. We use this to find the actual
+        // first thread from the linked list node contained by the first thread.
+        Target.TypeInfo type = _target.GetTypeInfo(DataType.Thread);
+        _threadLinkOffset = (ulong)type.Fields[nameof(Data.Thread.LinkNext)].Offset;
     }
 
     ThreadStoreData IThread.GetThreadStoreData()
     {
         Data.ThreadStore? threadStore;
-        if (!_target.ProcessedData.TryGet(_threadStoreAddr.Value, out threadStore))
+        if (!_target.ProcessedData.TryGet(_threadStoreAddr, out threadStore))
+        {
+            threadStore = new Data.ThreadStore(_target, _threadStoreAddr);
+
+            // Still okay if processed data is already registered by someone else
+            _ = _target.ProcessedData.TryRegister(_threadStoreAddr, threadStore);
+        }
+
+        return new ThreadStoreData(
+            threadStore.ThreadCount,
+            GetThreadFromLink(threadStore.FirstThreadLink),
+            _target.ReadGlobalPointer(Constants.Globals.FinalizerThread),
+            _target.ReadGlobalPointer(Constants.Globals.GCThread));
+    }
+
+    ThreadStoreCounts IThread.GetThreadCounts()
+    {
+        Data.ThreadStore? threadStore;
+        if (!_target.ProcessedData.TryGet(_threadStoreAddr, out threadStore))
         {
             threadStore = new Data.ThreadStore(_target, _threadStoreAddr);
 
             // Still okay if processed data is already registered by someone else
-            _ = _target.ProcessedData.TryRegister(_threadStoreAddr.Value, threadStore);
+            _ = _target.ProcessedData.TryRegister(_threadStoreAddr, threadStore);
+        }
+
+        return new ThreadStoreCounts(
+            threadStore.UnstartedCount,
+            threadStore.BackgroundCount,
+            threadStore.PendingCount,
+            threadStore.DeadCount);
+    }
+
+    ThreadData IThread.GetThreadData(TargetPointer threadPointer)
+    {
+        Data.Thread? thread;
+        if (!_target.ProcessedData.TryGet(threadPointer, out thread))
+        {
+            thread = new Data.Thread(_target, threadPointer);
+
+            // Still okay if processed data is already registered by someone else.
+            _ = _target.ProcessedData.TryRegister(threadPointer, thread);
         }
```

Comment:
I wish we could make something reusable for this pattern.  How wasteful (I guess primarily in terms of generated code size) would passing a closure be:

```csharp
   thread = _target.ProcessedData.GetOrAdd(threadPointer, static (target, addr) => new Data.Thread(target, addr));
```

```csharp
// DataCache
public T GetOrAdd<T>(TargetPointer addr, Func<Target,TargetPointer,T> ctor) {
  if (TryGet(addr, out T result) {
    return result;
  } else {
    T constructed = ctor(_target, addr);
    if (TryRegister(addr, constructed)) {
      return constructed;
    } else {
      bool found = TryGet (addr, out T result);
      Debug.Assert(found);
      return result;
    }
  }
}
```

Reply:
that's nice